rust/lance-index/src/vector/ivf/shuffler.rs

```diff
@@ -100,37 +118,52 @@ impl IvfShuffler {
 
         file_writer.finish().await?;
 
+        unsafe {
+            self.set_unsorted_buffers(vec![UNSORTED_BUFFER.into()]);
+        }
+
         Ok(())
     }
 
-    async fn total_batches(&self) -> Result<usize> {
-        let object_store = ObjectStore::local();
-        let path = self.output_dir.child(UNSORTED_BUFFER);
-        let reader = FileReader::try_new(&object_store, &path).await?;
-        Ok(reader.num_batches())
+    async fn total_batches(&self) -> Result<Vec<usize>> {
+        let mut total_batches = vec![];
+        for buffer in &self.unsorted_buffers {
+            let object_store = ObjectStore::local();
+            let path = self.output_dir.child(buffer.as_str());
+            let reader = FileReader::try_new(&object_store, &path).await?;
+            total_batches.push(reader.num_batches());
+        }
+        Ok(total_batches)
     }
 
-    async fn count_partition_size(&self, start: usize, end: usize) -> Result<Vec<u64>> {
+    async fn count_partition_size(&self, inputs: &Vec<ShuffleInput>) -> Result<Vec<u64>> {
         let object_store = ObjectStore::local();
-        let path = self.output_dir.child(UNSORTED_BUFFER);
-        let reader = FileReader::try_new(&object_store, &path).await?;
-
         let mut partition_sizes = vec![0; self.num_partitions as usize];
 
-        let lance_schema = reader
-            .schema()
-            .project(&[PART_ID_COLUMN])
-            .expect("part id should exist");
-
-        let mut stream = stream::iter(start..end)
-            .map(|i| reader.read_batch(i as i32, ReadBatchParams::RangeFull, &lance_schema))
-            .buffered(64);
-
-        while let Some(batch) = stream.next().await {
-            let batch = batch?;
-            let part_ids: &UInt32Array = batch.column(0).as_primitive();
-            for part_id in part_ids.values() {
-                partition_sizes[*part_id as usize] += 1;
+        for &ShuffleInput {
+            file_idx,
+            start,
+            end,
+        } in inputs
+        {
+            let file_name = &self.unsorted_buffers[file_idx];
+            let path = self.output_dir.child(file_name.as_str());
+            let reader = FileReader::try_new(&object_store, &path).await?;
+            let lance_schema = reader
+                .schema()
+                .project(&[PART_ID_COLUMN])
+                .expect("part id should exist");
+
+            let mut stream = stream::iter(start..end)
+                .map(|i| reader.read_batch(i as i32, ReadBatchParams::RangeFull, &lance_schema))
+                .buffered(16);
+
+            while let Some(batch) = stream.next().await {
```

Comment:
Foreach?

Reply:
we need to update `partition_sizes` `Fn` closures can't do that easily.